admin/github.go

```diff
@@ -216,15 +221,88 @@ func (s *Service) processGithubPush(ctx context.Context, event *github.PushEvent
 
 func (s *Service) processGithubInstallationEvent(ctx context.Context, event *github.InstallationEvent) error {
 	switch event.GetAction() {
-	case "created", "unsuspend", "suspend", "new_permissions_accepted":
+	case "created", "unsuspend", "new_permissions_accepted":
 		// TODO: Should we do anything for unsuspend?
-	case "deleted":
-	}
+	case "suspend", "deleted":
+		// the github installation ID will change if user re-installs the app deleting the project for now
+		installation := event.GetInstallation()
+		if installation == nil {
+			return fmt.Errorf("nil installation")
+		}
 
+		if err := s.deleteProjectsForInstallation(ctx, installation.GetID()); err != nil {
+			s.logger.Error("failed to delete project for installation", zap.Int64("installation_id", installation.GetID()), zap.Error(err))
+			return err
+		}
+	}
 	return nil
 }
 
 func (s *Service) processGithubInstallationRepositoriesEvent(ctx context.Context, event *github.InstallationRepositoriesEvent) error {
 	// We can access event.RepositoriesAdded and event.RepositoriesRemoved
+	switch event.GetAction() {
+	case "added":
+		// no handling as of now
+	case "removed":
+		var multiErr error
+		for _, repo := range event.RepositoriesRemoved {
+			if err := s.deleteProjectsForRepo(ctx, repo); err != nil {
+				multiErr = multierr.Combine(multiErr, err)
+				s.logger.Error("failed to delete project for repo", zap.String("repo", *repo.HTMLURL), zap.Error(err))
+			}
+		}
+		return multiErr
+	}
 	return nil
 }
+
+func (s *Service) deleteProjectsForInstallation(ctx context.Context, id int64) error {
+	// Find Rill project for installationID
+	projects, err := s.DB.FindProjectsByGithubInstallationID(ctx, id)
+	if err != nil {
+		if errors.Is(err, database.ErrNotFound) {
+			// App is removed from repo not currently deployed. Do nothing.
+			return nil
+		}
+		return err
+	}
+
+	var multiErr error
+	for _, p := range projects {
+		err := s.TeardownProject(ctx, p)
+		if err != nil {
+			if !errors.Is(err, database.ErrNotFound) {
+				multiErr = multierr.Combine(multiErr, err)
+				s.logger.Error("unable to delete project", zap.String("project_id", p.ID), zap.Error(err))
+			}
+			continue
+		}
+	}
+	return multiErr
+}
+
+func (s *Service) deleteProjectsForRepo(ctx context.Context, repo *github.Repository) error {
+	// Find Rill project matching the repo that was pushed to
+	githubURL := *repo.HTMLURL
+	projects, err := s.DB.FindProjectsByGithubURL(ctx, githubURL)
+	if err != nil {
+		if errors.Is(err, database.ErrNotFound) {
+			// App is removed from repo not currently deployed. Do nothing.
+			return nil
+		}
+		return err
+	}
+
+	var multiErr error
+	for _, p := range projects {
+		err := s.TeardownProject(ctx, p)
+		if err != nil {
+			if !errors.Is(err, database.ErrNotFound) {
+				multiErr = multierr.Combine(multiErr, err)
+				s.logger.Error("unable to delete project", zap.String("project_id", p.ID), zap.Error(err))
```

Comment:
This log is redundant because the caller is logging the returned multierr. You can add the `project_id` to the error message if you want it to get printed (something like `fmt.Errorf("unable to delete project %q: %w", p.ID, err)`)